sgl-kernel/csrc/sgl_diffusion/elementwise/timestep_embedding.cu

```diff
@@ -0,0 +1,161 @@
+/* Copyright 2025 SGLang Team. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
+#include <ATen/cuda/CUDAContext.h>
+#include <cuda_bf16.h>
+#include <cuda_fp16.h>
+#include <cuda_runtime.h>
+#include <math.h>
+#include <torch/all.h>
+
+#include <cassert>
+#include <cmath>
+
+template <typename T>
+__device__ float convert_to_float(T x) {
+  if constexpr (std::is_same_v<T, __half>) {
+    return __half2float(x);
+  } else if constexpr (std::is_same_v<T, __nv_bfloat16>) {
+    return __bfloat162float(x);
+  } else if constexpr (std::is_same_v<T, float>) {
+    return x;
+  } else {
+    // int, double
+    return static_cast<float>(x);
+  }
+}
+
+template <typename T>
+__device__ __nv_bfloat16 convert_to_bfloat16(T x) {
+  if constexpr (std::is_same_v<T, __nv_bfloat16>) {
+    return x;
+  } else if constexpr (std::is_same_v<T, __half>) {
+    return __float2bfloat16(__half2float(x));
+  } else if constexpr (std::is_same_v<T, float>) {
+    return __float2bfloat16(x);
+  } else {
+    return __float2bfloat16(static_cast<float>(x));
+  }
+}
+
+template <typename T>
+__device__ __half convert_to_float16(T x) {
+  if constexpr (std::is_same_v<T, __half>) {
+    return x;
+  } else if constexpr (std::is_same_v<T, __nv_bfloat16>) {
+    return __float2half(__bfloat162float(x));
+  } else if constexpr (std::is_same_v<T, float>) {
+    return __float2half(x);
+  } else {
+    return __float2half(static_cast<float>(x));
+  }
+}
+
+template <typename O, typename T>
+__device__ O cast_to(T x) {
+  if constexpr (std::is_same_v<O, float>) {
+    return convert_to_float(x);
+  } else if constexpr (std::is_same_v<O, __half>) {
+    return convert_to_float16(x);
+  } else if constexpr (std::is_same_v<O, __nv_bfloat16>) {
+    return convert_to_bfloat16(x);
+  } else {
+    return static_cast<O>(convert_to_float(x));
+  }
+}
+
+template <typename T_IN>
+__global__ void
+timestep_embedding_kernel(T_IN* t_ptr, float* output_ptr, int dim, float neg_log_max_period, int batch_size) {
+  // Get the timestep for this batch
+  int row_idx = blockIdx.x * blockDim.y + threadIdx.y;
+  if (row_idx >= batch_size) {
+    return;
+  }
+  float t_val = cast_to<float>(__ldg(&t_ptr[row_idx]));
```

Comment:
You can use [castToFloat and castFromFloat](https://github.com/sgl-project/sglang/blob/main/sgl-kernel/include/utils.h#L359-L367) and delete above dtype convert codes.